vehicle/src/Vehicle/Compile/Print/TypingError.hs

```diff
@@ -251,12 +249,17 @@ typeRestrictionError ctx (TypeRestrictionOrigin freeEnv (ident, p) sort typ) _ca
             <+> "to a supported type"
     }
   where
+    supportedTypesDescription = case sort of
+      Left _ -> "The possible valid types for" <+> quotePretty sort <+> "annotated declarations are:"
+      Right RestrictedRecord -> "All fields of a record annotated with" <+> quotePretty sort <+> "must have the same type and must be either:"
```

Comment:
To get a good error message here, I think you may need to case split earlier, and actually generate an entire fresh `problem` field depending on whether it is `Left` or `Right`.